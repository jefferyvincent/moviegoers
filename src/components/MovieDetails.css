.movie-details-body {
    display: flex;
}

.backBtn {
    cursor: pointer;
    text-decoration: underline;
    color: #00c8ff;
}

.movie-details-body img{
    margin-right: 2rem;
    max-width: 100%;
}

.user-scores {
    display: flex;
}

.user-score {
    font-size: small;
    margin-top: 1rem;
    margin-left: 1rem;
}

.rating {
    padding: 1rem;
}

.red {
    background-color: red;
    color: white;
}

.green {
    background-color: green;
    color: white;
}

@media only screen and (max-width: 904px) {
    .movie-details-body {
        flex-wrap: wrap;
        justify-content: center;
    }
}