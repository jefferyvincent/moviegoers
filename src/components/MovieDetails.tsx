import React, {useState, useEffect, Fragment} from 'react';
import axios from 'axios';

import Layout from './Layout';
import './MovieDetails.css';
import noImage from '../img/no_image.png';

interface MovieDetailsProps {
    match: any;
    history: any;
};

const MovieDetails:React.FC<MovieDetailsProps> = ({match, history}) => {
    const [data, setData] = useState();

    useEffect(() => {
        const fetchData = async () => {
            const result = await axios(
            `https://api.themoviedb.org/3/movie/${match.params.id}?api_key=438581175542f3ba5208b33d46f87151`,
            );
            setData(result.data);
        };
        fetchData();
    }, [match.params.id]);

    const goBack = ():void => {
        history.goBack();
    }

    const renderProductionCo = () => {
        const len = data.production_companies.length;
        if (len > 0) {
            return data.production_companies.map((item:any, i:number) =>{
                if (len === i + 1) {
                    return <span key={i}>{item.name}</span>;
                } else {
                    return <span key={i}>{item.name}, </span>;
                }
            });
        } else {
            return <span>No Data</span>
        }
    }

    const renderGenres = () => {
        const len = data.genres.length;
        if (len > 0) {
            return data.genres.map((item: any, i:number) => {
                if (len === i + 1) {
                    return <span key={i}>{item.name}</span>;
                } else {
                    return <span key={i}>{item.name}, </span>;
                }
            });
        } else {
            return <span>No Data</span>
        }
    }

    const renderVotingAverage = () => {
        let score = 'red';
        if (Math.round(parseInt(data.vote_average)) > 5) {
            score = 'green';
        }
        return (
            <div className="user-scores">
                <div className={"rating " + score}>{Math.round(parseInt(data.vote_average))}</div>
                <div className="user-score">Based on {data.vote_count} votes</div>
            </div>
        )
    }

    const renderOverview = () => {
        if (data.overview !== null) {
            return data.overview;
        } else {
            return <span>No Data</span>
        }
    }

    const renderRunTime = () => {
        if (data.runtime !== null) {
            return <span>{data.runtime} mins</span>;
        } else {
            return <span>No Data</span>
        }
    }

    const renderDate = () => {
        if (data.release_date !== null) {
        
            let d = new Date(data.release_date);

            const monthNames = ["January", "February", "March", "April", "May", "June",
            "July", "August", "September", "October", "November", "December"
            ];

            let date = monthNames[d.getUTCMonth()] +" " + d.getDate() + ", " + d.getFullYear();

            return <span>{date}</span>;
        } else {
            return <span>No Data</span>
        }
    }
    const renderResult = () => {
        if (data !== undefined) {
            if (data.hasOwnProperty('id')) {

                // image
                let img = noImage;
                if (data.poster_path !== null) {
                    img = "https://image.tmdb.org/t/p/w500/"+data.poster_path
                }

                return (
                    <Fragment>
                        <div data-testid="resolved" className="backBtn" onClick ={goBack}>Back</div>
                        <div className="movie-details">
                            <h2>{data.original_title}</h2>
                            <div className="movie-details-body">
                                <img src={img} alt={data.original_title} />
                                <div>
                                    {renderVotingAverage()}
                                    <p><strong>Release Date:</strong> {renderDate()}</p>
                                    <p><strong>Run time:</strong> {renderRunTime()}</p>
                                    <p><strong>Plot Summary:</strong> {renderOverview()}</p>
                                    <p><strong>Genres:</strong> {renderGenres()}</p>
                                    <p><strong>Production Co: </strong>{renderProductionCo()}</p>
                                </div>
                            </div>
                        </div>
                    </Fragment>
                ) 
            } else {
                return <div data-testid="loader" className="loading">loading...</div>
            }
        } else {
            return <div data-testid="loader" className="loading">loading...</div>
        }
    }

    return (
        <Layout>
            {renderResult()}
        </Layout>
    );
}

export default MovieDetails;