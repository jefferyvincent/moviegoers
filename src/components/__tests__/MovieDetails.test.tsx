import React from 'react';
import { shallow } from 'enzyme';
import {render } from 'react-testing-library';
import {HashRouter as Router} from "react-router-dom";
import MovieDetails from '../MovieDetails';

describe('<MovieDetails/>', () => {
  let wrapper:any;
  beforeEach(() => { wrapper = shallow(<MovieDetails match={{ params: { id: " 302349"} } } history={[]} />); });

  it("renders", () => {
    const { asFragment } = render(<Router><MovieDetails match={{ params: { id: " 302349"} } } history={[]} /></Router>);
    expect(asFragment()).toMatchSnapshot();
  });
  
  
  it('renders an loader', () => {
    expect(wrapper.find('div.loading').length).toEqual(1);
  });

  /* 
    Need to mock Ajax somehow here???? Nothing is is working and __mocks__ / axios.js is throwing me errors. 
    Reference:
    https://www.leighhalliday.com/mocking-axios-in-jest-testing-async-functions
    https://www.leighhalliday.com/async-axios-react-testing-library
  */
});