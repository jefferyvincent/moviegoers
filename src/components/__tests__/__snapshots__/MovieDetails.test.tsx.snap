// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieDetails/> renders 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <header
      class="app-header"
    >
      <div
        class="header-left"
      >
        <a
          href="#/"
          title="Ultimate Moviegoers Guide"
        >
          <h1
            class="header-logo"
          >
            <img
              alt="Ultimate Moviegoers Guide"
              src="logo.png"
            />
          </h1>
        </a>
      </div>
      <div
        class="header-right"
      >
        <div
          class="app-search"
        >
          <input
            autocomplete="off"
            autocorrect="off"
            placeholder="search it."
            spellcheck="false"
          />
          <img
            alt="Search"
            class="search-mag"
            src="mag.svg"
          />
        </div>
      </div>
    </header>
    <main
      class="content"
    >
      <div
        class="loading"
      >
        loading...
      </div>
    </main>
    <footer
      class="footer"
    >
      <hr />
      <div
        class="copyright"
      >
        Â© 2019 Ultimate Moviegoers Guide
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
